### PR TITLE
層数と次元数が可変の各種コンポーネントモデルを作成

### PR DESCRIPTION
**Pull request by seichi042I**
## 概要
#88 
各モジュールのlinearをLinearLayersに変更

## 変更内容
- src/models/components/controller.pyを変更
- src/models/components/observation_auto_encoder.pyを変更
- src/models/components/prior_encoder.pyを変更

## 影響範囲
上記のファイルをimportしているファイル
- dreamer.py
- agent.py
- world.py

## Submit前の確認項目

- [x] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [x] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [x] この PR で導入されたすべての変更点をリストアップしましたか?
- [x] PR を `pytest`または `make test`コマンドでローカルにテストしましたか?
- [x] `pre-commit run -a` または`make format`コマンドで `pre-commit` フックを実行しましたか?

## 補足
変更後，すべてのテストの動作確認済み

### CODE REVIEW

File: src/models/components/linear_layers.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            bias (bool):
+        """
+        super().__init__()
+        self.input_layer = torch.nn.Linear(input_size,hidden_size,bias=bias)
+
+        self.hidden_layers = torch.nn.Sequential()
+
+        for i in range(layers):
+            self.hidden_layers.add_module(f"linear_{i}",torch.nn.Linear(hidden_size,hidden_size,bias=bias))
+            self.hidden_layers.add_module(f"ReLU_{i}",torch.nn.ReLU())
```
Comment (Geson-anko) [inline src/models/components/linear_layers.py:31]: BatchNorm 1dも欲しいですね。総数を深くしたときに値の爆発を防いでくれます。

File: src/models/components/controller.py [hunk 2]
```diff
@@ -52,8 +53,18 @@ def __init__(
...
+            hidden_size=c_hidden_size * 2,
+            layers=3,
+            output_size=action_size,
+        )
+        self.fc_logs = LinearLayers(
+            input_size=c_hidden_size,
+            hidden_size=c_hidden_size * 2,
+            layers=3,
+            output_size=action_size,
+        )
```
Comment (Geson-anko) [inline src/models/components/controller.py:67]: fc_meanとfc_logsは計算量の削減のためにただの線形層でお願いします
その前に`fc_hiddens`等の名前で`LinearLayer`を定義しておいて、そこから分岐させる方法を取りましょう。

LinearLayersのhidden_sizeもコンストラクタの引数から直接指定できるようにしましょう。レイヤーの数も同様です。
```suggestion
        self.fc_hiddens = LinearLayers(c_hidden_size, stacked_linear_hidden_size)
        
        self.fc_mean = nn.Linear(stacked_linear_hidden_size, action_size)
        self.fc_logs = nn.Linear(stacked_linear_hidden_size, action_size)
```

File: src/models/components/linear_layers.py (hunk 1 above)
Comment (Geson-anko) [inline src/models/components/linear_layers.py:31]: `append`で追加できます。  
```suggestion
            self.hidden_layers.append(torch.nn.Linear(hidden_size,hidden_size,bias=bias))
            self.hidden_layers.append(torch.nn.ReLU())
```

File: src/models/components/observation_auto_encoder.py [hunk 3]
```diff
@@ -26,8 +27,18 @@ def __init__(
...
+            hidden_size=(state_size + hidden_size) * 2,
+            layers=3,
+            output_size=state_size,
+        )
+        self.fc_logs = LinearLayers(
+            input_size=state_size + hidden_size,
+            hidden_size=(state_size + hidden_size) * 2,
+            layers=3,
+            output_size=state_size,
+        )
```
Comment (Geson-anko) [inline src/models/components/observation_auto_encoder.py:41]: ここもfc_hidden_layersとしてまとめて、最後に`fc_mean`や`fc_logs`を通す形をとってください。そしてLinearLayersの中間層のサイズをコンストラクタで指定できるようにしてください。レイヤーの数も同様です。

File: src/models/components/observation_auto_encoder.py [hunk 4]
```diff
@@ -105,7 +116,12 @@ def __init__(
             bias=conv_bias,
         )
 
-        self.voc_decoder = torch.nn.Linear(self.decoder.idim, voc_state_size)
+        self.voc_decoder = LinearLayers(
+            input_size=self.decoder.idim,
+            hidden_size=self.decoder.idim * 2,
+            layers=3,
+            output_size=voc_state_size,
+        )
```
Comment (Geson-anko) [inline src/models/components/observation_auto_encoder.py:124]: ここもhidden_sizeとlayersをコンストラクタで指定できるようにしましょう。

File: src/models/components/prior_encoder.py [hunk 5]
```diff
@@ -24,8 +25,12 @@ def __init__(self, hidden_dim: int, state_dim: int, min_stddev: float = 0.1) ->
...
         assert state_dim > 0, "state_dim must be greater than 0"
 
-        self.fc_to_mean = nn.Linear(hidden_dim, state_dim)
-        self.fc_to_stddev = nn.Linear(hidden_dim, state_dim)
+        self.fc_to_mean = LinearLayers(
+            input_size=hidden_dim, hidden_size=hidden_dim * 2, layers=3, output_size=state_dim
+        )
+        self.fc_to_stddev = LinearLayers(
+            input_size=hidden_dim, hidden_size=hidden_dim * 2, layers=3, output_size=state_dim
+        )
```
Comment (Geson-anko) [inline src/models/components/prior_encoder.py:33]: ここもcontrollerのレビューと同じくです。

**Review by Geson-anko — CHANGES_REQUESTED** [review]
お疲れ様です。いい感じに実装出来てきましたね。
LinearLayersを実装した各コンポーネントは`src/models/components/stacked_linear/`に新しく作成して、controller, prior_encoder, observation_auto_encoderの変更を破棄してください。(Revert commit) 現在の実験を中断しないためです。(transitionの修正はOKです。ありがとうございます)

試験期間で動けなかったのでめっちゃ助かります！ありがとうございます！

**Review by cehl-kurage — COMMENTED** [review]
追加で修正のリクエストはありません

File: src/models/components/linear_layers.py [hunk 6]
```diff
@@ -15,17 +17,16 @@ def __init__(
...
+        self.hidden_layers = nn.Sequential()
 
-        for i in range(layers):
-            self.hidden_layers.add_module(
-                f"linear_{i}", torch.nn.Linear(hidden_size, hidden_size, bias=bias)
-            )
-            self.hidden_layers.add_module(f"ReLU_{i}", torch.nn.ReLU())
+        for _ in range(layers):
+            self.hidden_layers.append(nn.Linear(hidden_size, hidden_size, bias=bias))
+            self.hidden_layers.append(nn.LayerNorm(hidden_size))
```
Comment (cehl-kurage) [inline src/models/components/linear_layers.py:26]: レビューでは`BatchNorm1d`を追加することになってましたが、この変更の理由は何でしょう？

Comment (Geson-anko) [reply src/models/components/linear_layers.py:26]: @cehl-kurage これは RNNに適用する関係でBatchNormが期待通り動作しないことが想定されるためです。  
BatchNormは時間軸も含めたミニバッチ全体に適用したいのですが、RNNは性質上時間方向を別々に処理するので、各タイムステップごとに正則化されてしまいます。こういった観点からLayer Normalizationを使用しました。シフトとスケールがデータごとに常に0, 1になる効果は回帰問題においてデメリットがありますが、中間出力とする分には問題ないと判断しました。

**Review by Geson-anko — APPROVED** [review]
コードを大きく改変したためレビュアーから辞退します
